text/0000-generic-assert.md

```diff
@@ -11,23 +11,27 @@ Make the `assert!` macro generic to all expressions, and extend the readability
 # Motivation
 [motivation]: #motivation
 
-While clippy warns about `assert!` usage that should be replaced by `assert_eq!`, it's quite annoying to migrating around.
+While clippy warns about `assert!` usage that should be replaced by `assert_eq!`, it's quite annoying to migrate around.
 
 Unit test frameworks like [Catch](https://github.com/philsquared/Catch) for C++ does cool message printing already by using macros.
 
 # Detailed design
 [design]: #detailed-design
 
-We're going to parse AST and break up them by operators (excluding `.` (dot, member access operator)). Function calls and bracket surrounded blocks are considered as one block and don't get split further.
+We're going to parse AST and break up them by operators (excluding `.` (dot, member access operator)). Function calls and bracket surrounded blocks are considered as one block and don't get expanded. The exact expanding rules should be determined when implemented, but an example is provided for reference.
 
 On assertion failure, the expression itself is stringified, and another line with intermediate values are printed out. The values should be printed with `Debug`, and a plain text fallback if the following conditions fail:
 - the type doesn't implement `Debug`.
 - the operator is non-comparison (those in `std::ops`) and the type (may also be a reference) doesn't implement `Copy`.
 
 To make sure that there's no side effects involved (e.g. running `next()` twice on `Iterator`), each value should be stored as temporaries and dumped on assertion failure.
 
+The new assert messages are likely to generate longer code, and it may be simplified for release builds (if benchmarks confirm the slowdown).
```

Comment:
There's really no reason to include this functionality in release builds at all. You don't want the increased code size, nor this information, in released software.

Reply:
One of Rust's advantage is that you can get backtrace properly even in release builds. Basically, if you don't know how the assertion failed, it would be almost impossible to diagnose the crash.

Reply:
Right, and I can appreciate that, but there can also be other constraints in play such as in embedded systems. You might want to disable all the extras and just handle errors that occur by restarting.

Reply:
I'll add that in commercial software it's also desirable to hide all information about implementation details.

Reply:
@repax: You could easily redefine your own `assert!` to hide every detail.

```rust
#[cfg(not(debug_assertions))]  // <------
macro_rules! assert {
    ($e:expr) => {
        if !$e {
            panic!("assertion failure");
        }
    };
    ($e:expr, $($rest:tt)*) => {
        assert!($e)
    }
}

fn main() {
    let a = 2;
    let b = 2;
    let c = 5;
    assert!(a + b == c, "note: failed assertion {} + {} == {}", a, b, c);
}
```

Reply:
@kennytm: Why not just make this a compile-time option? Sometimes you want this debugging convenience, often times not. In designing features like this I think it's important to take into account the breadth of contexts in which rust code may operate in the wild.

Reply:
Define a cargo feature and use `#[cfg(feature = "redacted_assert")]` then.

Reply:
My macro probably wouldn't apply to 3rd party crates, so this doesn't quite solve the bloat issue.